lib/plugins/rotate_asg_instances/asg_checks.rb

```diff
@@ -0,0 +1,67 @@
+module Moonshot
+  # All the checks related methods for the rotation plugin should add here.
+  module ASGChecks
+    def try_to_ssh(raise_exception: false)
+      execute_ssh
+    rescue StandardError
+      if raise_exception
+        raise "SSH failed. exit status: #{@command_response.exitstatus} " \
+              "error: #{@command_response.output}"
+      else
+        false
+      end
+    end
+
+    # Runs the ssh command.
+    #
+    # @param instance [Aws::EC2::Instance] (Target instance)
+    # @param command [String] (Command to execute)
+    # @param timeout_connection [Boolean] (Timeout the connection)
+    # @return [Struct] (:output and :exitstatus)
+    #   Output of the command executed
+    def run_ssh_command(instance, command, timeout_connection: true)
+      remote_cmd =
+        build_command(instance, command, timeout_connection)
+      Moonshot::SSHForkExecutor.new.run(remote_cmd)
+    end
+
+    private
+
+    def print_ssh_debug?
+      ENV['PRINT_SSH_DEBUG'] ? true : false
+    end
+
+    # Builds the command for ssh execution.
+    #
+    # @param instance [Aws::EC2::Instance] (Target instance)
+    # @param command [String] (Command to execute)
+    # @param timeout_connection [Boolean] (Timeout the connection)
+    # @return [String] (ssh command)
+    #   Command to be executed
+    def build_command(instance, command, timeout_connection)
+      remote_cmd = "#{@ssh_user}@#{instance.public_dns_name}"
+      command_options = 'ssh '
+      options = ['UserKnownHostsFile=/dev/null', 'StrictHostKeyChecking=no']
+      command_options += '-v ' if print_ssh_debug?
+      command_options += '-o ConnectTimeout=10' if timeout_connection
+      "#{command_options} -o #{options.join(' -o ')} #{remote_cmd} #{command}"
+    end
+
+    def execute_ssh
+      Retriable.retriable(base_interval: 5, tries: 3) do
+        @command_response = run_ssh_command(first_ec2_instance, '/bin/true')
+        raise StandardError.new, 'Validation failed.' unless
+          @command_response.exitstatus.zero?
+        @ilog.info(@command_response.output.split(/\r\n/)) if print_ssh_debug?
+        @command_response.exitstatus.zero?
+      end
+    end
+
+    def first_ec2_instance
+      first_instance = asg.instances.first
+      return if first_instance.nil?
+
+      Aws::EC2::Instance.new(id: first_instance.instance_id)
+    end
```

Comment:
I think we should use the classes that moonshot have to ssh into the instance, here we are just adding a few options that I think we don't want to add, like ignoring the strict host key check. Another point is that the builder command class already has code to select an instance and we don't need to duplicate this in this class.
And then maybe we can just call the Moonshot ssh classes from ASG rotate plugin class.

Reply:
Makes sense